subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/detection/DefaultTestExecuter.java

```diff
@@ -149,4 +162,61 @@ private int getMaxParallelForks(JvmTestExecutionSpec testExecutionSpec) {
         }
         return maxParallelForks;
     }
+
+    /**
+     * Create a classpath or modulePath, as a set of files, given both the files provided by the test spec and a set of
+     * modules to potentially load from the Gradle distribution. For each additional module to load, we check if the
+     * test files contains a jar which satisfies the module-to-load. If the test files do not contain a satisfactory
+     * jar, the given additional module is loaded from the Gradle distribution.
+     *
+     * TODO: This process is performed on a best-effort basis by matching against test file jar names. In the future
+     * may want to consider updating this code to take dependency-management into account, which can provide
+     * a more accurate view of {@code testFiles} that does not depend on string regex matching.
+     *
+     * @param testFiles A set of jars, as given from a {@link JvmTestExecutionSpec}'s classpath or modulePath.
+     * @param additionalModules The names of any additional modules to potentially load from the Gradle distribution.
+     *
+     * @return A set of files representing the constructed classpath or modulePath.
+     */
+    private Set<File> pathWithAdditionalJars(Iterable<? extends File> testFiles, Set<? extends TestFramework.DistributionModule> additionalModules) {
+        Set<File> outputFiles = new LinkedHashSet<File>();
+
+        Set<? extends TestFramework.DistributionModule> mutableAdditionalModules = new HashSet<TestFramework.DistributionModule>(additionalModules);
+        for (File f : testFiles) {
+            // Loop through each additional jar from the distribution. If the test classpath provides
+            // this jar, don't load it from the distribution.
+            Iterator<? extends TestFramework.DistributionModule> it = mutableAdditionalModules.iterator();
+            while (it.hasNext()) {
+                TestFramework.DistributionModule module = it.next();
+                if (module.getFileNameMatcher().matcher(f.getName()).matches()) {
```

Comment:
If we're only matching on the jar names, a quicker way to implement this detection might be to first sort each list of jars, then step through each in order to remove the common names.  This would get rid of the nested loops and possibly be faster for large lists of jars.  Could also try something fancier with building a `Map` to the elements of `additionalMutableModules`, keyed by their names.  

I just think the potential for many dozens or even hundreds of jars on these paths exists and it's worth caring about how this operation slows at scale.

Reply:
The list of modules to load from the distribution currently has a max size of 3, as defined by the `TestFramework`s. We do not plan on expanding this list any further. So, currently this runs in 3n time, where n is the number of jars in the test-defined classpath (`testFiles`). Sorting would first cost n*log(n) plus any further detection logic on top of that. 

We can't strictly use string equality here since we only load modules from the distribution by name, and we are not aware of the version. We need to perform some more complex checking here, as necessitated the regular expressions. 

I did have performance in mind when writing this. We are using pre-compiled regular expressions, which are only compiled once when the `TestFramework` class is loaded. Further, we only iterate over `testFiles` once, an operation which was already required.